File: 98.validate-binary-search-tree.js
// @leet start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
	return helper(root, -Infinity, Infinity);
};
var helper = function(root, left, right) {
	if (!root) return true;
	if (root.val <= left || root.val >= right) return false;
	return (
		helper(root.left, left, root.val) && helper(root.right, root.val, right)
	);
};
// @leet end

File: 98.validate-binary-search-tree.java
// @leet start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public boolean isValidBST(TreeNode root) {
    return dfs(root, Long.MIN_VALUE, Long.MAX_VALUE);
  }
  public boolean dfs(TreeNode root, long min, long high) {
    if (root == null)
      return true;
    if (root.val <= min || root.val >= high)
      return false;
    return dfs(root.left, min, root.val) && dfs(root.right, root.val, high);
  }
}
// @leet end
